### PR TITLE
Add vector slice to fuzzer

### PR DESCRIPTION
**Pull request by Yuhta**
Differential Revision: D38801564

### CODE REVIEW

File: velox/vector/fuzzer/VectorFuzzer.h [hunk 1]
```diff
@@ -106,6 +106,7 @@ class VectorFuzzer {
   // Wraps `vector` using a randomized indices vector, returning a
   // DictionaryVector.
   VectorPtr fuzzDictionary(const VectorPtr& vector);
+  VectorPtr fuzzDictionary(const VectorPtr& vector, vector_size_t size);
```
Comment (xiaoxmeng) [inline velox/vector/fuzzer/VectorFuzzer.h:109]: Can you comment on size?

File: velox/vector/fuzzer/VectorFuzzer.cpp [hunk 2]
```diff
@@ -364,16 +383,23 @@ VectorPtr VectorFuzzer::fuzzComplex(
...
 VectorPtr VectorFuzzer::fuzzDictionary(const VectorPtr& vector) {
+  return fuzzDictionary(vector, vector->size());
+}
+
+VectorPtr VectorFuzzer::fuzzDictionary(
+    const VectorPtr& vector,
+    vector_size_t size) {
   const size_t vectorSize = vector->size();
-  BufferPtr indices = AlignedBuffer::allocate<vector_size_t>(vectorSize, pool_);
+  VELOX_DCHECK(!(vectorSize == 0 && size > 0));
```
Comment (xiaoxmeng) [inline velox/vector/fuzzer/VectorFuzzer.cpp:393]: What you want to check here? I am assuming size should not be smaller than vector size given the current usage?

File: velox/vector/fuzzer/VectorFuzzer.cpp [hunk 3]
```diff
@@ -263,9 +269,22 @@ VectorFuzzer::fuzz(const TypePtr& type, vector_size_t size, bool flatEncoding) {
...
+    size = outputSize;
+    if (size > 0 && coinToss(0.05)) {
+      size += folly::Random::rand32(8, rng_);
+    }
+    vector = fuzzDictionary(vector, size);
+    if (size > outputSize) {
+      auto offset = folly::Random::rand32(size - outputSize + 1, rng_);
+      vector = vector->slice(offset, outputSize);
+    }
   }
```
Comment (xiaoxmeng) [inline velox/vector/fuzzer/VectorFuzzer.cpp:288]: Add VELOX_CHECK here vector size is outputSize?

File: velox/vector/fuzzer/VectorFuzzer.cpp [hunk 4]
```diff
@@ -243,6 +243,12 @@ VectorPtr
 VectorFuzzer::fuzz(const TypePtr& type, vector_size_t size, bool flatEncoding) {
   VectorPtr vector;
 
+  vector_size_t outputSize = size;
```
Comment (xiaoxmeng) [inline velox/vector/fuzzer/VectorFuzzer.cpp:246]: nit: I'd use a new local var actualSize= size and change actualSize instead of function input size.

**Review by xiaoxmeng — COMMENTED** [review]
Looks good. Some minors.

File: velox/vector/fuzzer/VectorFuzzer.cpp (hunk 2 above)
Comment (Yuhta) [reply velox/vector/fuzzer/VectorFuzzer.cpp:393]: No `size` is the number of indices and irrelevant to the size of underlying vector.  The only corner case here is we cannot build a non-empty dictionary out of an empty underlying vector.

**Review by xiaoxmeng — APPROVED** [review]
LGTM. Thanks for the change!

**Pull request closed without merging**